### PR TITLE
Introduce CGAL::Default for the TDS of 2D constrained triangulations

### PR DESCRIPTION
**Pull request by afabri**
[Small feature "Use Default"](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Use_Default)

This PR makes several typedefs in examples obsolete.

### CODE REVIEW

**Comment by lrineau** [comment]
TODO: documentation

**Comment by lrineau** [comment]
Will be tested in [CGAL-4.9-Ic-5][5].

[5]: https://cgal.geometryfactory.com/CGAL/testsuite/results-4.9-Ic-5.shtml

**Comment by lrineau** [comment]
I have an error, in `examples/Barycentric_coordinates_2/Terrain_height_modeling.cpp`:
```
/home/lrineau/Git/cgal/Barycentric_coordinates_2/examples/Barycentric_coordinates_2/Terrain_height_modeling.cpp:79:90:   required from here
/home/lrineau/Git/cgal/Triangulation_2/include/CGAL/Constrained_triangulation_2.h:562:43: error: cannot call member function ‘CGAL::Constrained_triangulation_2<Gt, Tds_, Itag_>::Vertex_handle CGAL::Constrained_triangulation_2<Gt, Tds_, Itag_>::insert(const Point&, CGAL::Constrained_triangulation_2<Gt, Tds_, Itag_>::Locate_type, CGAL::Constrained_triangulation_2<Gt, Tds_, Itag_>::Face_handle, int) [with Gt = CGAL::Projection_traits_xy_3<CGAL::Epick>; Tds_ = CGAL::Triangulation_data_structure_2<CGAL::Triangulation_vertex_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick> >, CGAL::Delaunay_mesh_face_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick> > >; Itag_ = CGAL::No_intersection_tag; CGAL::Constrained_triangulation_2<Gt, Tds_, Itag_>::Vertex_handle = CGAL::internal::CC_iterator<CGAL::Compact_container<CGAL::Triangulation_vertex_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick>, CGAL::Triangulation_ds_vertex_base_2<CGAL::Triangulation_data_structure_2<CGAL::Triangulation_vertex_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick> >, CGAL::Delaunay_mesh_face_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick> > > > >, CGAL::Default, CGAL::Default, CGAL::Default>, false>; CGAL::Constrained_triangulation_2<Gt, Tds_, Itag_>::Point = CGAL::Point_3<CGAL::Epick>; CGAL::Constrained_triangulation_2<Gt, Tds_, Itag_>::Locate_type = CGAL::Triangulation_2<CGAL::Projection_traits_xy_3<CGAL::Epick>, CGAL::Triangulation_data_structure_2<CGAL::Triangulation_vertex_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick> >, CGAL::Delaunay_mesh_face_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick> > > >::Locate_type; CGAL::Constrained_triangulation_2<Gt, Tds_, Itag_>::Face_handle = CGAL::internal::CC_iterator<CGAL::Compact_container<CGAL::Delaunay_mesh_face_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick>, CGAL::Constrained_Delaunay_triangulation_face_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick>, CGAL::Constrained_triangulation_face_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick>, CGAL::Triangulation_face_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick>, CGAL::Triangulation_ds_face_base_2<CGAL::Triangulation_data_structure_2<CGAL::Triangulation_vertex_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick> >, CGAL::Delaunay_mesh_face_base_2<CGAL::Projection_traits_xy_3<CGAL::Epick> > > > > > > >, CGAL::Default, CGAL::Default, CGAL::Default>, false>]’ without object
   return Constrained_triangulation::insert(a,lt,loc,li);
                                           ^
```

**Comment by afabri** [comment]
fixed. I had to put back the real default types.

**Comment by lrineau** [comment]
https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-4.9-Ic-5/output/Triangulation_2/classCGAL_1_1Constrained__Delaunay__triangulation__2.html

Missing space between "_2" and "or":
> Tds	must be a model of TriangulationDataStructure_2or Default.

Anyway, it would better turn this PR and the other one into a small feature.

**Comment by afabri** [comment]
I don't get the point. This does exactly what the 3D triangulation does  for years.

**Comment by lrineau** [comment]
Internally. The use of `Default` was never documented. For example, the page 
http://doc.cgal.org/latest/Triangulation_3/classCGAL_1_1Delaunay__triangulation__3.html does not contain any mention of `Default`.

In my opinion, it would be better if the modifications to the documentation would be reviewed by a small feature.

**Comment by afabri** [comment]
[Small Feature](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Use_Default)

File: Triangulation_2/include/CGAL/Constrained_triangulation_2.h [hunk 1]
```diff
@@ -40,16 +40,25 @@ struct Exact_predicates_tag{}; // to be used with filtered exact number
...
 
 public:
+  typedef typename Default::Get<Tds_, Triangulation_data_structure_2 <
+                                        Triangulation_vertex_base_2<Gt>,
+                                        Constrained_triangulation_face_base_2<Gt> > >::type Tds;
+  typedef typename Default::Get<Itag_, No_intersection_tag>::type Itag;
+
   typedef Triangulation_2<Gt,Tds> Triangulation;
-  typedef Constrained_triangulation_2<Gt,Tds,Itag>  Constrained_triangulation;
+  typedef Constrained_triangulation_2<Gt,Tds_,Itag>  Constrained_triangulation;
```
Comment (mglisse) [inline Triangulation_2/include/CGAL/Constrained_triangulation_2.h]: Why did Tds gain an underscore but not Itag here? It seems to cause your issues with Default as default parameter. I don't see much point to this typedef though, its only use seems questionable.

Comment (afabri) [inline Triangulation_2/include/CGAL/Constrained_triangulation_2.h]: Because there is later  a  `typedef typename Ctr::Tds Tds;`

Comment (mglisse) [inline Triangulation_2/include/CGAL/Constrained_triangulation_2.h]: I wasn't asking why you renamed the template parameter, I was asking specifically about line 61 why you forgot to update `Itag` to its new name `Itag_` in the typedef for `Constrained_triangulation`.

**Comment by afabri** [comment]
@mglisse   Can you have a look.

**Comment by mglisse** [comment]
LGTM. Is there another issue that prevents from doing the same thing (use Default as default parameter) in `Constrained_Delaunay_triangulation_2`, for consistency?

**Comment by lrineau** [comment]
Is this PR ready for testing?

**Comment by lrineau** [comment]
@afabri?
> Is this PR ready for testing?

**Comment by afabri** [comment]
yes, please.

**Pull request merged**